avocado/utils/configure_network.py

```diff
@@ -10,193 +10,323 @@
+        ip = ip_interface("{}/{}".format(ipaddr, netmask))
+        cmd = 'ip addr add {} dev {}'.format(ip.compressed,
+                                             self.name)
+        try:
+            _run_command(cmd, self.remote_session, sudo=True)
+        except Exception as ex:
+            raise NWException("Failed to add address {}".format(ex))
 
-def unset_ip(interface):
-    """
-    Gets interface name unassigns the IP to the interface
-    """
-    if distro.detect().name == 'rhel':
-        conf_file = "/etc/sysconfig/network-scripts/ifcfg-%s" % interface
+    def bring_down(self):
+        """Utility used to Bring down interface """
 
-    if distro.detect().name == 'SuSE':
-        conf_file = "/etc/sysconfig/network/ifcfg-%s" % interface
+        cmd = "ifdown {}".format(self.name)
+        try:
+            _run_command(cmd, self.remote_session, sudo=True)
+        except Exception as ex:
+            raise NWException("ifdown fails: %s" % ex)
 
-    cmd = "ifdown %s" % interface
-    try:
-        process.system(cmd, sudo=True)
-        shutil.move(conf_file+".backup", conf_file)
-        return True
-    except Exception as ex:
-        raise NWException("ifdown fails: %s" % ex)
+    def bring_up(self):
+        """Utility used to Bring up interface"""
 
+        cmd = "ifup {}".format(self.name)
+        try:
+            _run_command(cmd, self.remote_session, sudo=True)
+        except Exception as ex:
+            raise NWException("ifup fails: %s" % ex)
 
-def ping_check(interface, peer_ip, count, option=None, flood=False):
-    """
-    Checks if the ping to peer works.
-    """
-    cmd = "ping -I %s %s -c %s" % (interface, peer_ip, count)
-    if flood is True:
-        cmd = "%s -f" % cmd
-    elif option is not None:
-        cmd = "%s %s" % (cmd, option)
-    if process.system(cmd, shell=True, verbose=True,
-                      ignore_status=True) != 0:
-        return False
-    return True
-
-
-def set_mtu_host(interface, mtu):
-    """
-    Set MTU size in host interface
-    """
-    cmd = "ip link set %s mtu %s" % (interface, mtu)
-    try:
-        process.system(cmd, shell=True)
-    except process.CmdError as ex:
-        raise NWException("MTU size can not be set: %s" % ex)
-    try:
-        cmd = "ip add show %s" % interface
-        mtuvalue = process.system_output(cmd, shell=True).decode("utf-8") \
-                                                         .split()[4]
-        if mtuvalue == mtu:
+    def is_link_up(self):
+        """
+        Checks if the interface link is up
+        :return: True if the interface's link up, False otherwise.
+        """
+        if self.get_link_state() in ['up', 'UP']:
             return True
-    except Exception as ex:  # pylint: disable=W0703
-        log.error("setting MTU value in host failed: %s", ex)
-    return False
+        else:
+            return False
 
+    def is_remote(self):
+        if self.remote_session:
```

Comment:
@apahim I don't think that I got the problem in this case.

`if self.remote_session`,  will test against an empty string and empty lists too. I'm not only checking if is "not None" but also doing some basic checks too.

On another hand, with `if self.remote_session is not None` you just check if is not None.

An even better if would be, checking if the class type of self.remote_session, to make sure that `ssh.Session`.

Could you please elaborate here?

Reply:
This check is checking if self.remote_session was set and for that you check if it is not `None`. If you want to validate the `self.remote_session`, you should do it properly.

Reply:
Checking if it's an instance of `ssh.Session` would make a good validation indeed.
To be clear, I'd personally not do any validation at all, but my point is, as I said: if you want to validate, do it properly :)

Reply:
@apahim yes I understand, but then we should define what is "was set". Because for me, `remote_session = ''` in this case, is an invalid set.

If we would like to check if something is true/valid, the proper way, IMO, is to do `if self.remote_session`. Because it will call `.__nonzero__() / `__bool__()`:

> Called to implement truth value testing and the built-in operation bool(); should return False or True. When this method is not defined, __len__() is called, if it is defined, and the object is considered true if its result is nonzero. If a class defines neither __len__() nor __bool__(), all its instances are considered true.

Each class might have a different way to determine if something is "set" or not. And this is why I'm using only `if`.

But, for sure, I agree that if we are going to validate, let's do it right. I will change this to check for the instance type instead.

Reply:
I can create a valid object, that is an instance of `ssh.Session`, and that evaluates to `False` when you call bool(obj). On the other hand, remote_session = 'foobar' is invalid and you do nothing about it. So, I insist, please don't mix validation with checking if a parameter was provided.

Reply:
Maybe the 'instead" on my sentence gave the impression that I'm was going to do *only* the instance check. That was not my intention. My idea is to do this:

```python
if self.remote_session and isinstance(self.remote_session, ssh.Session):
```
The first will check for "the parameter was provided", this is not validation. And the second one is the validation itself.

IMO, `if self.remote_session is not None` is not a proper validation (because empty strings will fail here) and *also not* a proper check if something was set, because of the same reason ("was set" is relative to the class).

So, when I do: `if something:` I'm not doing validations here. I'm just doing the "was set" (or a "exists") check, that is, IMO, a more complete check than `is not None`.

Reply:
Yep, we clearly disagree :)

Reply:
@apahim I would love to understand why we disagree or at least understand if this, in particular, is something that will block this PR. If so, I have no problem in using something that I do not agree, I will be happy to just move on.

Please, keep in mind that: I understand your point of not mix checks with validation. I agree with it. My only point is that I was not doing both.

The point here is what is the better way to check If a "bag of popcorn" exists?
a) `if bag_of_popcorn:` (that will check if the bag is empty or not) or `if bag_of_popcorn is not None`?

I'm not checking if the popcorn is old (expired) here. I'm not talking about validations.

Reply:
No, I'd not block a PR for that.

Your new example, about the `bag_of_popcorn` has nothing to do with the case here, so I'm not commenting on it.

About the pull request code, the default value is `None`. When you do `if self.remote_session:`, you are not only checking whether it was set to something, you're also checking whether that something evaluates to `True` when you call `bool(self.remote_session)`. While it might look like a good idea, because you want an empty string, an empty list, an integer with value `0`,... all that to be ruled out, conceptually it's the wrong thing to do, since the `bool()` evaluation to `True` is not the protocol that one should implement for a valid object. For example, considering that your condition is:

```
        if self.remote_session:
```

`self.remote_session = 0` fails the check, but `self.remote_session = 1` passes the check. `self.remote_session = ''` fails the check, but `self.remote_session = ' '` passes the check. That's inconsistent, since the objects passing the check are not valid.  

If you now ALSO check the inheritance with:

```
        if self.remote_session and isinstance(self.remote_session, ssh.Session):
```

An instance of this class will be accepted:
```
class MyCustomRemote1(ssh.Session):
    pass
```

While an instance of this one will not:
```
class MyCustomRemote2(ssh.Session):
    def __bool__(self):
        return False
```

The reason why `bool(MyCustomRemote2)` evaluates to `False` is none of your concern, so why do you care? Instead, you should just say: if there's something, anything, I will use it. And the way to do that is `if self.remote_session is not None`.

That said, since you are also intending to validate the object inheritance, that alone should suffice, meaning that your `if` would be just `if isinstance(self.remote_session, ssh.Session):`, but that should probably go in the class `__init__()` and raise an attribute error when the check fails.

I've written a lot, but I'm afraid I have said nothing new and I hate repeating myself, so I'm sorry about that.

Reply:
> About the pull request code, the default value is `None`. When you do `if self.remote_session:`, you are not only checking whether it was set to something, you're also checking whether that something evaluates to `True` when you call `bool(self.remote_session)`. While it might look like a good idea, because you want an empty string, an empty list, an integer with value `0`,... all that to be ruled out, conceptually it's the wrong thing to do, since the `bool()` evaluation to `True` is not the protocol that one should implement for a valid object. 

I see, but that is my point: I'm not trying to find if the object is _valid_ or not here, this is not a validation. That is why I described the "bag of popcorn" example, to exercise the question of existence (truth test) not validity.

> The reason why bool(MyCustomRemote2) evaluates to False is none of your concern, so why do you care?

Because when I think "if something exists", I'm thinking "is something != 0", like the "bag of popcorn" example. If you say: "please, grab me a bag of popcorn", and I bring you an empty bag, not sure if you would be happy with that. (or a "glass of water").

`bool()` will perform the standard "Truth Value Testing" (truth testing procedure)[1], that will check not only for `__bool__` but also for empty sequences and collections in some cases. When I'm trying to find the answer to the question "is this exists?" Mentally, I'm trying to run the "Truth Value Testing" and check if this something is != 0.

If you are deliberately changing this `__bool__()`  to return true, then you are also changing the "Truth Value Testing".

What I could think out of this discussion is: there are 3 questions:

 1. Is something set?
 2. Is something not empty? (the python truth test)
 3. Is something valid?

The proposed code is doing 2 because I believe that the "python truth test" here is more suitable. 


> That said, since you are also intending to validate the object inheritance, that alone should suffice, meaning that your if would be just if isinstance(self.remote_session, ssh.Session):, but that should probably go in the class __init__() and raise an attribute error when the check fails.

I agree that only `isinstance()` will be `enough` here to check if is a remote session (and I will follow this approach). But I don't think that raising an exception or add this to the `__init__` is needed because this is something optional, it will add unnecessary lines to the `__init__`.

> I've written a lot, but I'm afraid I have said nothing new and I hate repeating myself, so I'm sorry about that.

No worries, we are here to share ideas and IMO this discussion is healthy and needed if we would like to have a good "Contributor's Guide".

So, I'm sorry too if I'm being too repetitive. Also, I don't know if this is already a consensus with the team and was broadly discussed before. I apologize in advance if that was the case.

[1] - https://docs.python.org/3/library/stdtypes.html#truth

Reply:
Go for it. As I said, it's a non-blocker. Thanks for the discussion :)